Add tests that objects are created in correct realms

ricea opened the pull request:
When Javascript in multiple iframes interacts it becomes important to
distinguish which objects belong to which realm. Add tests that verify
that objects emitted by TextEncoder and TextDecoder streams were created
in the correct realm.

ricea committed:
Add tests that objects are created in correct realms

When Javascript in multiple iframes interacts it becomes important to
distinguish which objects belong to which realm. Add tests that verify
that objects emitted by TextEncoder and TextDecoder streams were created
in the correct realm.

tyoshino commented on tests/realms.html:
```diff
@@ -0,0 +1,341 @@
+function createRealm() {
+  let iframe = document.createElement('iframe');
+  iframe.srcdoc = `<!doctype html>
+<script src="resources/text-encode-transform.js">${'<'}/script>
```
what's the reason that we use the expression for '<'?

ricea replied:
It hides </script> from the HTML parser. There's probably a better way to do it but I haven't worked out what it is yet.

tyoshino replied:
ahhh, ok

ricea replied:
I tried storing the whole tag in a variable. Please take a look to see if it's clearer.

tyoshino replied:
Great. Thanks!

tyoshino commented on tests/realms.html:
```diff
@@ -0,0 +1,341 @@
+    assert_equals(result.constructor, constructorRealm.Object,
+                  'result should be in constructor realm');
+    // Strings are always in the current realm, so no point in checking.
+   }, 'the result object when write is called with a pending ' +
```
indent

tyoshino reviewed (commented):
Still trying to understand what's done here!

ricea committed:
Clarify escaped script end tag

Also fix indentation.

tyoshino commented on tests/realms.html:
```diff
@@ -0,0 +1,342 @@
+                  'readable should be in method realm');
+    assert_equals(writable.constructor, methodRealm.WritableStream,
+                  'writable should be in method realm');
+  }, `${classname}\'s readable and writable attributes should come from the ` +
```
`\'s` doesn't need escaping?

ricea replied:
Fixed.

tyoshino commented on tests/realms.html:
```diff
@@ -0,0 +1,342 @@
+  return realm[myId];
+}
+
+// Construct an object in constructedRealm and copy it into readRealm and
```
Construct -> Constructs
for consistency

ricea replied:
Fixed.

tyoshino commented on tests/realms.html:
```diff
@@ -0,0 +1,342 @@
+    await writePromise;
+    assert_equals(result.constructor, constructorRealm.Object,
+                  'result should be in constructor realm');
+    // Strings are always in the current realm, so no point in checking.
```
Sorry. I'm not familiar with this point. Could you give me some pointer to the normative document to understand this?

ricea replied:
I've made the comment more specific of the exact behaviour I observe. @domenic, you know your way around the ECMAScript spec, do you know where this behaviour is defined?

domenic replied:
Strings as primitives do not have a realm at all.

However, they get implicitly converted from lowercase-S "string" primitives to an uppercase-S `String` object if you try to access properties of them, e.g. `"foo".constructor` or similar.

This implicit conversion always uses the current realm. It happens in https://tc39.github.io/ecma262/#sec-toobject but it doesn't explicitly state the realm, I think because in ES current is such an assumed default that nobody even thought to specify it.

tyoshino replied:
Got it. Thanks Domenic!

tyoshino commented on tests/realms.html:
```diff
@@ -0,0 +1,342 @@
+      assert_equals(err.constructor, constructorRealm.TypeError,
+                    'read TypeError should come from constructor realm');
+    }
+  }, 'TypeError for chunk if the wrong type should come from constructor ' +
```
if -> with?

ricea replied:
Fixed, thanks.

tyoshino commented on tests/realms.html:
```diff
@@ -0,0 +1,342 @@
+    // Read first to relieve backpressure.
+    const readPromise = readInReadRealm(objId);
+    // Write an invalid chunk.
+    const invalidBytesId = id();
```
not used?

ricea replied:
Removed, thanks.

ricea committed:
Fix comments and messages

Make the comment about why we don't perform realm checks on strings more
specific.

Remove unused "invalidBytesId" variable.

Various textual fixes.

tyoshino commented:
f9ab81ee97a34075917f2dd8381e745697458ba0 lgtm
